PictureInPictureKotlin/app/src/main/java/com/example/android/pictureinpicture/MainViewModel.kt

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (C) 2021 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.example.android.pictureinpicture
+
+import android.os.SystemClock
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.map
+import androidx.lifecycle.viewModelScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.android.awaitFrame
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.launch
+import kotlin.time.Duration
+
+class MainViewModel: ViewModel() {
+
+    private var job: Job? = null
+
+    private var startUptimeMillis = SystemClock.uptimeMillis()
+    private val timeMillis = MutableLiveData(0L)
+
+    private val _started = MutableLiveData(false)
+
+    val started: LiveData<Boolean> = _started
+    val time = timeMillis.map { millis ->
+        val minutes = millis / 1000 / 60
+        val m = minutes.toString().padStart(2, '0')
+        val seconds = (millis / 1000) % 60
+        val s = seconds.toString().padStart(2, '0')
+        val hundredths = (millis % 1000) / 10
+        val h = hundredths.toString().padStart(2, '0')
+        "$m:$s:$h"
+    }
+
+    /**
+     * Starts the stopwatch if it is not yet started, or pauses it if it is already started.
+     */
+    fun startOrPause() {
+        if (_started.value == true) {
+            _started.value = false
+            job?.cancel()
+        } else {
+            _started.value = true
+            job = viewModelScope.launch {
+                startUptimeMillis = SystemClock.uptimeMillis() - (timeMillis.value ?: 0L)
+                while (isActive) {
+                    timeMillis.value = SystemClock.uptimeMillis() - startUptimeMillis
+                    awaitFrame()
+                }
```

Comment:
I wonder if it might be helpful to put this into a separate `suspend fun`? I completely missed the start of the `launch` block when going through it the first time.
More of a 'nit' thing would be to maybe add a comment about `awaitFrame()`?

Reply:
Done